// amf-bot-framework/src/renderers/choice-renderer.js
import builder from 'botbuilder';

/**
 * Class for the renderer of the Choice cards.
 * These cards are the ones that gives to the user several options to choose.
 * The options can be shown as buttons, numbered list or inline numbered list.
 */
class ChoiceRenderer {

  /**
   * Class for the renderer of the Choice cards.
   * These cards are the ones that gives to the user several options to
   * choose.
   * The options can be shown as buttons, numbered list or inline
   * numbered list.
   *
   * @constructor
   * @param bot { Object } Parent bot.
   */
  constructor(bot) {
    this.bot = bot;
  }

  /**
   * Main render method of the class. Renders the card into Microsoft Bot
   * Framework objects, and send them to the front.
   *
   * @param session { Object } Chat session.
   * @param card { Object } Card instance.
   * @param cb { Function } Callback function.
   */
  render(session, card, cb) {
    this.bot.domain.renderObj(session, card, function(error, card) {
      if (error) {
        return cb(error);
      }
      session.dialogData.lastPrompt = card;
      var options = '';
      for (var i = 0; i < card.options.length; i++) {
        if (i > 0) {
          options += '|';
        }
        options += card.options[i].text;
      }
      var resultcard = {
        text: card.text,
        type: card.subtype ? card.subtype : 'auto',
        src: card,
        options: options
      };
      session.dialogData.lastChoice = card;
      var style;
      if (card.type) {
        style = builder.ListStyle[card];
      }
      if (style) {
        builder.Prompts.choice(session, card.text, options,
          {listStyle: style});
      } else {
        builder.Prompts.choice(session, card.text, options);
      }
      return cb(null, resultcard);
    });
  }
}

export default ChoiceRenderer;
