// amf-bot-framework/examples/demobot/common/core/bot-cards.js
var _ = require('lodash');

var cards = {
    invalidOption: _.clone({
        type: 'text',
        text: 'Invalid option. Please select a valid one.'
    }),
    youChoosed: _.clone({
        type: 'text',
        text: 'You choosed %s'
    }),
    healthCard: _.clone({
        type: 'hero',
        title: 'My AXA Health',
        text: 'MyAXAhealth offers you instant access to your Health policies '+
          'and employee benefits with AXA anytime, and anywhere for '+
          'your convenience.',
        image: '{{{ root.url }}}/images/health.png',
        tap: {
            type: 'openUrl',
            url: '{{{ root.url }}}/images/health.png'
        },
        buttons: [
            { 
                type: 'openUrl',
                url: 'https://play.google.com/store/apps/details?'+
                  'id=com.axa.wellbe.lu',
                title: 'Store'
            }
        ]
        
    }),
    menuIntegrationsText: _.clone({
        type: 'choice',
        subtype: 'button',
        text: 'Which Integration do you want to test?',
        isMenu: true,
        options: [
            { tag: 'int-analyze-image', text: 'Analyze Image'},
            { tag: 'int-thumbnail', text: 'Thumbnail'},
            { tag: 'int-spell-check', text: 'Spell Check'},
            { tag: 'int-jira', text: 'Jira'},
            { tag: 'int-weather', text: 'Weather'},
            { tag: 'quit', text: 'Quit'}
        ] 
    }),
    analyzeImageText: _.clone({
        type: 'prompt',
        subtype: 'attachment',
        text: 'I can analyze images. Send me an image and I\'ll analyze it.'
    }),
    thumbnailText: _.clone({
        type: 'prompt',
        subtype: 'attachment',
        text: 'I can generate thumbnails. Send me an image.'
    }),
    inputWidth: _.clone({
        type: 'prompt',
        subtype: 'number',
        text: 'Enter the width',
        scope: 'dialog',
        name: 'width'
    }),
    inputHeight: _.clone({
        type: 'prompt',
        subtype: 'number',
        text: 'Enter the height',
        scope: 'dialog',
        name: 'height'
    }),
    spellCheckText: _.clone({
        type: 'prompt',
        subtype: 'text',
        text: 'Enter some text and I will check the spell',
        scope: 'dialog',
        name: 'text'
    }),
    weatherText: _.clone({
        type: 'prompt',
        subtype: 'text',
        text: 'Enter the location',
        scope: 'dialog',
        name: 'location'
    }),
    naturalText: _.clone({
        type: 'text',
        text: 'Tell me something using your own language',
    }),
    dontUnderstand: _.clone({
        type: 'text',
        text: 'I don\'t understand'
    }),
    multiriskIntro: _.clone({
        type: 'text',
        text: 'Welcome to the multirisk calculator'
    }),    
    inputAge: _.clone({
        type: 'prompt',
        subtype: 'number',
        text: 'Enter your age',
        scope: 'user',
        name: 'age'
    }),
    inputWeight: _.clone({
        type: 'prompt',
        subtype: 'number',
        text: 'Enter your weight',
        scope: 'dialog',
        name: 'weight'
    }),
    inputCountry: _.clone({
        type: 'prompt',
        subtype: 'text',
        text: 'Enter your country',
        scope: 'user',
        name: 'country'
    }),
    inputDrinks: _.clone({
        type: 'prompt',
        subtype: 'number',
        text: 'How many alcohol drinks you consume per week?',
        scope: 'user',
        name: 'drinks'
    }),
    wholelifeCard: _.clone({
        type: 'hero',
        title: 'Whole Life',
        image: '{{{ root.url }}}/images/wholelife1.png',
        tap: {
            type: 'openUrl',
            url: '{{{ root.url }}}/images/wholelife1.png'
        },
        buttons: [
            { 
                type: 'openUrl',
                url: 'https://play.google.com/store/apps/'+
                  'details?id=com.axa.wellbe.lu',
                title: 'Information'
            },
            { 
                type: 'openUrl',
                url: 'https://play.google.com/store/apps/details?'+
                  'id=com.axa.wellbe.lu',
                title: 'select'
            }
        ]
    }),
    termlifeCard: _.clone({
        type: 'hero',
        title: 'Term Life',
        image: '{{{ root.url }}}/images/termlife1.png',
        tap: {
            type: 'openUrl',
            url: '{{{ root.url }}}/images/termlife1.png'
        },
        buttons: [
            { 
                type: 'openUrl',
                url: 'https://play.google.com/store/apps/details?'+
                  'id=com.axa.wellbe.lu',
                title: 'Information'
            },
            { 
                type: 'openUrl',
                url: 'https://play.google.com/store/apps/details?'+
                  'id=com.axa.wellbe.lu',
                title: 'select'
            }
        ]
    }),
    criticalIllnessCard: _.clone({
        type: 'hero',
        title: 'Critical Illness',
        image: '{{{ root.url }}}/images/criticalillness1.png',
        tap: {
            type: 'openUrl',
            url: '{{{ root.url }}}/images/criticalillness1.png'
        },
        buttons: [
            { 
                type: 'openUrl',
                url: 'https://play.google.com/store/apps/details?'+
                  'id=com.axa.wellbe.lu',
                title: 'Information'
            },
            { 
                type: 'openUrl',
                url: 'https://play.google.com/store/apps/details?'+
                  'id=com.axa.wellbe.lu',
                title: 'select'
            }
        ]
    })
};

module.exports = cards;
